clients/venice-admin-tool/src/main/java/com/linkedin/venice/TopicMessageFinder.java

```diff
@@ -0,0 +1,148 @@
+import java.util.Date;
+import java.util.Map;
+import java.util.Properties;
+import org.apache.avro.Schema;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.clients.consumer.ConsumerRecords;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.clients.consumer.OffsetAndTimestamp;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+/**
+ * This class is used to find messages for a given key from a specified topic.
+ */
+public class TopicMessageFinder {
+  private static final Logger LOGGER = LogManager.getLogger(TopicMessageFinder.class);
+
+  public static void find(
+      ControllerClient controllerClient,
+      Properties consumerProps,
+      String topic,
+      String keyString,
+      long startTimestamp,
+      long endTimestamp,
+      long progressInterval) {
+    String storeName;
+    int version = -1;
+    if (Version.isVersionTopic(topic)) {
+      storeName = Version.parseStoreFromKafkaTopicName(topic);
+      version = Version.parseVersionFromKafkaTopicName(topic);
+    } else {
+      storeName = Version.parseStoreFromRealTimeTopic(topic);
+    }
+    // fetch key schema
+    String keySchemaStr = controllerClient.getKeySchema(storeName).getSchemaStr();
+    LOGGER.info("The key schema for store: {} : {}", storeName, keySchemaStr);
+    StoreInfo storeInfo = controllerClient.getStore(storeName).getStore();
+    int partitionCount = storeInfo.getPartitionCount();
+    // Parse key string and figure out the right partition
+    byte[] serializedKey = serializeKey(keyString, keySchemaStr);
+    if (version != -1) {
+      if (storeInfo.getVersion(version).isPresent()) {
+        if (storeInfo.getVersion(version).get().isChunkingEnabled()) {
+          serializedKey = new KeyWithChunkingSuffixSerializer().serializeNonChunkedKey(serializedKey);
+        }
+        partitionCount = storeInfo.getVersion(version).get().getPartitionCount();
+      } else {
+        throw new RuntimeException("Couldn't find version: " + version + " from store: " + storeName);
+      }
+    }
+    if (partitionCount == 0) {
+      throw new VeniceException("Invalid partition count: " + partitionCount);
+    }
+    LOGGER.info("Got partition count: {}", partitionCount);
+
+    int assignedPartition = new DefaultVenicePartitioner().getPartitionId(serializedKey, partitionCount);
+    LOGGER.info("Assigned partition: {} for key: {}", assignedPartition, keyString);
+
+    TopicPartition topicPartition = new TopicPartition(topic, assignedPartition);
+    long startOffset;
+    long endOffset;
+    consumerProps.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class);
+    consumerProps.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class);
```

Comment:
Why do we not use the same consumer/`consumerProps` for fetching offsets and consuming records?

Reply:
Discussed offline
> The ApacheKafkaConsumer doesn't expose a way to fetch offsets by time, but it is possible to use the raw consumer for polling recrods.

Reply:
Talked to @nisargthakkar offline, and he is fine to merge it as is since the perf is not concern and `ApacheKafkaConsumer` is easier to use for record polling.